### PR TITLE
Automate iPXE host installation for SL Micro 6.1

### PR DESCRIPTION
**Pull request by waynechen55**
* **This** pull request aims to support automated SL Micro 6.1 iPXE installation by using PXE-capable kernel/initrd and pre-built disk image (indicated by ```rd.kiwi.install.image=MIRROR_HTTP/HDD_1```) which then will be deployed onto specified disk device (indicated by ```rd.kiwi.oem.installdevice=/dev/disk/by-id/INSTALL_DISK_WWN```) and configured by using ignition/combustion (indicated by ```FIRST_BOOT_CONFIG```).

* **Several** new subroutines in ```lib/utils.pm```:
  * ```is_disk_image``` indicates pre-built disk image will used for OS installation.
  * ```is_ipxe_with_disk_image```  indicates OS installation is performed with iPXE and disk image.

* **New** test module ```jeos/prepare_firstboot_config.pm```:
  * **Aims** to generate ignition or combustion configuration file based on template ```config.ign.ARCH.ep``` or ```script.ARCH.ep```.
  * **Setting** ```FIRSTBOOT_CONFIG_DIR``` provides flexibility of using customized folder to store configuration templates.
  * **Subroutine** ```prepare_profile``` does the actual job of expanding variables (like ```SERIALCONSOLE```) and uploading final configuration file.
  * **Save** addresses of configuration files to settings ```IGNITION_URL``` and ```COMBUSTION_URL```.
  
* **Changes** made to ```tests/installation/ipxe_install.pm```:
  * **Move** part of adding extra kernel command line outside ```set_bootscript``` and into a new subroutine ```set_bootscript_cmdline_extra``` to avoid inflated ```set_bootscript``` subroutine.
  * **Introduce** ```BOOTLOADER_ROOT_DEVICE``` and ```BOOTLOADER_INITRD``` which indicates whether root device and specified initrd are needed at boot time for UEFI machines. If they are not needed, set ```BOOTLOADER_ROOT_DEVICE=0``` and ```BOOTLOADER_INITRD=0```. If customized ones are needed, set ```BOOTLOADER_ROOT_DEVICE=/dev/xxx``` and ```BOOTLOADER_INITRD=initrd_addr``` instead of default ```root=/dev/ram0``` and ```initrd=initrd```. Not using ```BOOTLOADER_ROOT_DEVICE``` and ```BOOTLOADER_INITRD``` means using default settings.
  * **New** subroutine ```set_bootscript_image_config``` adds boot parameters for pre-built disk image on kernel command line to achieve non-interactive mode installation, which is called in ```set_bootscript_cmdline_extra```. Setting ```rd.kiwi.oem.installdevice``` is only added if ```FIRST_BOOT_CONFIG``` is being used, which means full automation, and ```INSTALL_DISK_WWN``` will be set to ```/dev/sda``` or ```/dev/vda``` if it not given.
  * **New** subroutine ```set_bootscript_firstboot_config``` adds boot parameters for first boot configuration on kernel command line to complete automated configuration, which is called in ```set_bootscript_cmdline_extra```.
  * **Only** needles ```load-linux-kernel``` and ```load-initrd``` need to be matched for OS installation with disk image. 

* **Changes** made to ```tests/microos/selfinstall.pm```:
  * **For** installation with specified disk device, there will be no prompt confirmation of disk device, only dialogues of writing into and verifying disk device are shown up. Corresponding needles ```slem-selfinstall-write-drive``` and ```slem-selfinstall-verify-drive``` will be matched.
  * **Due** to test runs under various circumstances, ```check_screen('slem-selfinstall-verify-drive')``` will be used after ```assert_screen('slem-selfinstall-write-drive')``` to avoid unnecessary failures.
  
* **New** ignition (```config.ign.x86_64.ep``` and ```config.ign.aarch64.ep```) and combustion (```script.x86_64.ep``` and ```script.aarch64.ep```) configuration templates, which are based on architecture to better facilitate first boot configuration process.  These templates need to be expanded, because there are different serial consoles and etc being used for different machines, by executing ```jeos/prepare_firstboot_config.pm``` at the beginning of test run. Final configuration files are served on URL (```IGNITION_URL``` and ```COMBUSTION_URL```) and will be provided on kernel command line directly.

* **Add** new schedule file ```schedule/virt_autotest/install_guest_on_slem_kvm_host_ipxe_install.yaml``` for iPXE installation and rename existing one to ```schedule/virt_autotest/install_guest_on_slem_kvm_host_usb_install.yaml```.

* **Trivial** changes are made to legacy ```combustion/script``` file to take ```GRUB_TERMINAL_INPUT``` and ```GRUB_TERMINAL_OUTPUT``` into consideration as well.

* **Verification Runs:**
  * [15sp7 on 15sp7 kvm](https://openqa.suse.de/tests/15398680)
  * [15sp7 on 15sp7 xen](https://openqa.suse.de/tests/15398681) 
  * [15sp6 on 15sp6 kvm](https://openqa.suse.de/tests/15291447)
  * [15sp6 on 15sp6 xen](https://openqa.suse.de/tests/15291448)
  * [15sp6 on 12sp5 kvm](https://openqa.suse.de/tests/15274281)
  * [15sp6 on 12sp5 xen](https://openqa.suse.de/tests/15281228)
  * [15sp6 on 15sp5 kvm](https://openqa.suse.de/tests/15274292)
  * [15sp6 on 15sp5 xen](https://openqa.suse.de/tests/15274295)
  * [sl-micro 6.1 on sl-micro 6.1 host](https://openqa.suse.de/tests/15397757)
  * [sles15sp6 on sl-micro 6.1 host](https://openqa.suse.de/tests/15370301)
  * [ipxe x86_64 host installation](http://10.200.140.9/tests/15)
  * [ipxe aarch64 host installation](http://10.200.140.9/tests/26)
  * In case you can not see job on my openQA server
openqa_aarch64_ipxe_install_test_run_01
openqa_aarch64_ipxe_install_test_run_02

* **Needles:** 
  * Needle slem-selfinstall-write-drive
slem-selfinstall-write-drive
  * Needle slem-selfinstall-verify-drive 
slem-selfinstall-verify-drive

### CODE REVIEW

File: data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0 [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+# Configure grub2 in the system
+grub2_config
+
+# Keyboard
+systemd-firstboot --force --keymap=us
+
+# Registration
+if ! which SUSEConnect > /dev/null 2>&1; then
+    zypper --non-interactive install suseconnect-ng
```
Comment (mloviska) [inline data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0:28]: Why is this needed? Is there an image without suseconnect?

File: lib/utils.pm [hunk 2]
```diff
@@ -3136,4 +3139,43 @@ sub enable_console_kernel_log {
...
+
+=head2 is_disk_image
+
+ is_disk_image;
+
+Identify whether test runs with linux disk image built by kiwi or similar programs
+=cut
+
+sub is_disk_image {
+    return 1 if (get_var('HDD_1'));
```
Comment (mloviska) [inline lib/utils.pm:3165]: Many tests have HDD_1 defined, but that does not necessarily mean the variable represents a bootable image

Comment (waynechen55) [reply lib/utils.pm:3165]: It does not have to bootable I think, as long as they are used for installation and they are pre-built ones.

File: data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0 (hunk 1 above)
Comment (waynechen55) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0:28]: I am not sure. I think it would be better to add this.

Comment (mloviska) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0:28]: If the system is not already registered, then the zypper command should fail anyway.

File: lib/utils.pm (hunk 2 above)
Comment (mloviska) [reply lib/utils.pm:3165]: It can be used even for an empty drive. The subroutine in this form is ambiguous IMHO

Comment (waynechen55) [reply lib/utils.pm:3165]: > It can be used even for an empty drive. The subroutine in this form is ambiguous IMHO

Could you help share the job link ?

File: data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0 (hunk 1 above)
Comment (waynechen55) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0:28]: I will take this into consideration and take action if necessary.

Comment (waynechen55) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0]: I confirmed that even installing without repo returns successfully although no package will be installed. But I think you are right, so I removed this part.

**Comment by waynechen55** [comment]
@alice-suse @guoxuguang @Julie-CAO @RoyCai7 @nanzhg @tbaev Please have a look.

File: lib/utils.pm (hunk 2 above)
Comment (waynechen55) [reply lib/utils.pm]: I added ```is_sle_micro``` to make it more strict. It can also be extended in the future.

File: data/virt_autotest/host_unattended_installation_files/combustion/script.x86_64.ttyS1 [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+# combustion: network
+
+grub2_config() {
+    file=/etc/default/grub
+    sed -i 's/^\s*GRUB_TIMEOUT=.*$/GRUB_TIMEOUT=15/g' $file
+    sed -i -r 's/^(\s|#)*GRUB_TERMINAL=.*$/GRUB_TERMINAL="gfxterm console"/g' $file
+    sed -i -r 's/^(\s|#)*GRUB_TERMINAL_INPUT=.*$/GRUB_TERMINAL_INPUT="gfxterm console"/g' $file
+    sed -i -r 's/^(\s|#)*GRUB_TERMINAL_OUTPUT=.*$/GRUB_TERMINAL_OUTPUT="gfxterm console"/g' $file
+    sed -i 's/console=tty0//g' $file
+    sed -i 's/console=tty[^ ]*/console=ttyS1,115200/g' $file
```
Comment (alice-suse) [inline data/virt_autotest/host_unattended_installation_files/combustion/script.x86_64.ttyS1:11]: same here.
Next step, we will transfer more x86 HWs to uefi and used for sl micro test. Among them, serial console are not the same.

**Comment by alice-suse** [comment]
@czerw  This PR makes lots of changes to ipxe_install. Would you please have a look too?

**Comment by alice-suse** [comment]
@waynechen55 Great PR to make pxe installation for SL Micro work! As you know, currently there is a "pxe boot support" new jira feature for SL Micro 6.1 ongoing, have you got any clue/confirmation if the current pxe usage in the PR would be persistent after that jira feature implemented?

File: data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0 [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    sed -i -r 's/^(\s|#)*GRUB_TERMINAL_OUTPUT=.*$/GRUB_TERMINAL_OUTPUT="gfxterm console"/g' $file
+    sed -i 's/console=tty0//g' $file
+    sed -i 's/console=tty[^ ]*/console=ttyAMA0,115200/g' $file
+    echo "DEBUG: now $file is,"
+    cat $file
+    grub2-mkconfig -o /boot/grub2/grub.cfg
+}
+
+# Redirect output to the console
+exec > >(exec tee -a /dev/ttyAMA0) 2>&1
```
[the remark this answers is not in the document]
Comment (waynechen55) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0:18]: First boot configuration file provided and used directly on URL. It is not configurable and it is not necessary.

File: data/virt_autotest/host_unattended_installation_files/combustion/script.x86_64.ttyS1 (hunk 3 above)
Comment (waynechen55) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.x86_64.ttyS1:11]: First boot configuration file provided and used directly on URL. It is not configurable and it is not necessary. Then you create corresponding files for other serial consoles.

**Comment by alice-suse** [comment]
> [sl-micro 6.1 on sl-micro 6.1 host](https://openqa.suse.de/tests/15281992)
> [ipxe x86_64 host installation link1](http://10.202.64.26/tests/3121) [link2](http://10.202.96.18/tests/3111)
> [ipxe aarch64 host installation link1](http://10.202.64.26/tests/3119) [link2](http://10.202.96.18/tests/3112)

I can't open above links... Any hints?

**Comment by waynechen55** [comment]
> @waynechen55 Great PR to make pxe installation for SL Micro work! As you know, currently there is a "pxe boot support" new jira feature for SL Micro 6.1 ongoing, have you got any clue/confirmation if the current pxe usage in the PR would be persistent after that jira feature implemented?

The kernel command line settings are persistent. Files structure is configurable through MIRROR_HTTP, HDD_1 and etc. If changes have to be made afterwards, I think effort will be low.

**Comment by waynechen55** [comment]
> > [sl-micro 6.1 on sl-micro 6.1 host](https://openqa.suse.de/tests/15281992)
> > [ipxe x86_64 host installation link1](http://10.202.64.26/tests/3121) [link2](http://10.202.96.18/tests/3111)
> > [ipxe aarch64 host installation link1](http://10.202.64.26/tests/3119) [link2](http://10.202.96.18/tests/3112)
> 
> I can't open above links... Any hints?

Try again. If still no access, might be blocked by "new network change".

File: data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0 (hunk 4 above)
[the remark this answers is not in the document]
Comment (alice-suse) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0:18]: It's similar case with AUTOYAST files -- direct URL access in pxe. So it's doable, I think. It's not a big deal. You can think about it and then decide. I just give my suggestion :).

**Comment by czerw** [comment]
Please run verification on following generic jobs for aarch64 https://openqa.suse.de/tests/15292982 and x86_64 https://openqa.suse.de/tests/15293796

File: lib/Utils/Backends.pm [hunk 5]
```diff
@@ -262,7 +262,7 @@ sub is_generalhw { check_var('BACKEND', 'generalhw'); }
 #The ssh connection timeout is counted as seconds
 sub set_ssh_console_timeout {
     my ($sshd_config_file, $sshd_timeout) = @_;
-    $sshd_config_file //= '/etc/ssh/sshd_config';
+    $sshd_config_file = '/etc/ssh/sshd_config' if (!$sshd_config_file or script_run("ls $sshd_config_file") != 0);
```
Comment (czerw) [inline lib/Utils/Backends.pm:265]: It would be worth to comment why we need to do such things, which in my opinion doesn't look nice.

File: tests/installation/ipxe_install.pm [hunk 6]
```diff
@@ -88,41 +92,7 @@ sub set_bootscript {
...
-        $serial_dev = get_var('SERIALDEV', 'ttyS1');
-        $cmdline_extra .= " console=$serial_dev,115200 ";
-    }
-
-    # Extra options for virtualization tests with ipmi backend
-    $cmdline_extra .= " Y2DEBUG=1 linuxrc.log=/dev/$serial_dev linuxrc.core=/dev/$serial_dev linuxrc.debug=4,trace ";
-    $cmdline_extra .= " reboot_timeout=" . get_var('REBOOT_TIMEOUT', 0) . ' '
-      unless (is_leap('<15.2') || is_sle('<15-SP2'));
-    $cmdline_extra .= get_var('EXTRABOOTPARAMS', '');
+    my $cmdline_extra = set_bootscript_cmdline_extra();
```
Comment (czerw) [inline tests/installation/ipxe_install.pm:95]: Why it was moved to function? It is used only once.

Comment (czerw) [reply tests/installation/ipxe_install.pm:95]: Also is hard to what has changed as it was moved to different place.

File: tests/installation/ipxe_install.pm [hunk 7]
```diff
@@ -203,6 +273,11 @@ sub run {
 
     select_console 'sol', await_console => 0;
 
+    if (is_disk_image) {
+        check_screen([qw(load-linux-kernel load-initrd)], 120 / get_var('TIMEOUT_SCALE', 1));
```
Comment (czerw) [inline tests/installation/ipxe_install.pm:277]: What is the point of check- screen if we are not validating screen it at all?

File: tests/installation/ipxe_install.pm [hunk 8]
```diff
@@ -58,24 +58,28 @@ sub set_pxe_boot {
 
 sub set_bootscript {
     my $host = get_required_var('SUT_IP');
+    my $distri = is_sle_micro('>=6.1') ? "SL-Micro" : uc get_required_var('DISTRI');
```
Comment (czerw) [inline tests/installation/ipxe_install.pm:61]: Why do we need `uc` now?

File: tests/installation/ipxe_install.pm (hunk 6 above)
Comment (waynechen55) [reply tests/installation/ipxe_install.pm:95]: It is not in a different place. The new subroutine ```set_bootscript_cmdline_extra``` is in the same module. I think it would be better to wrap these things up, otherwise the ``set_bootscript``` will continue to expand in a way that does not look tidy. It is already a very lengthy subroutine.

File: tests/installation/ipxe_install.pm (hunk 7 above)
Comment (waynechen55) [reply tests/installation/ipxe_install.pm:277]: linx/initrd loading always being checked in this module for other installations, so I just keep it. I think it worth checking because linux/initrd loading is the first step or beginning of ipxe installation.

File: lib/Utils/Backends.pm (hunk 5 above)
Comment (waynechen55) [reply lib/Utils/Backends.pm:265]: Comment can be added.

File: data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0 (hunk 4 above)
[the remark this answers is not in the document]
Comment (waynechen55) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.aarch64.ttyAMA0]: Done

File: data/virt_autotest/host_unattended_installation_files/combustion/script.x86_64.ttyS1 (hunk 3 above)
Comment (waynechen55) [reply data/virt_autotest/host_unattended_installation_files/combustion/script.x86_64.ttyS1]: Done

**Comment by waynechen55** [comment]
> Please run verification on following generic jobs for aarch64 https://openqa.suse.de/tests/15292982 and x86_64 https://openqa.suse.de/tests/15293796

Please check them out:
https://openqa.suse.de/tests/15360145
https://openqa.suse.de/tests/15360146

**Comment by alice-suse** [comment]
@czerw It seems all comments are resolved. If no further suggestion, we plan to merge it tomorrow EOB. 
@waynechen55 I notice you have been keeping updating the PR in recent 2 days. My suggestion is that if it is code rework based on comments, let's include them in this PR. Otherwise better to open new PRs for further improvements, so that this PR can be merged without one more round of new review.

**Comment by waynechen55** [comment]
> @czerw It seems all comments are resolved. If no further suggestion, we plan to merge it tomorrow EOB. @waynechen55 I notice you have been keeping updating the PR in recent 2 days. My suggestion is that if it is code rework based on comments, let's include them in this PR. Otherwise better to open new PRs for further improvements, so that this PR can be merged without one more round of new review.

I will let you know when it is ready to merge.

**Comment by frankenmichl** [comment]
Please re-trigger the verification runs again before working when you're done with all changes

**Comment by alice-suse** [comment]
> > [sl-micro 6.1 on sl-micro 6.1 host](https://openqa.suse.de/tests/15281992)
> > [ipxe x86_64 host installation link1](http://10.202.64.26/tests/3121) [link2](http://10.202.96.18/tests/3111)
> > [ipxe aarch64 host installation link1](http://10.202.64.26/tests/3119) [link2](http://10.202.96.18/tests/3112)
> 
> I can't open above links... Any hints?

Finally, I can open them now.

Some question about  http://10.200.131.12:8666/ipxe/. Hope you can help explain. @waynechen55 
- what is this machine, 10.200.131.12? Do we use it in official test, or switch to official pxe server of OSD?
- I notice that there is no build info for the files under the directory. Where are they from? In official test, will they be updated per build? Which team shall do that?

**Comment by alice-suse** [comment]
> > > > [sl-micro 6.1 on sl-micro 6.1 host](https://openqa.suse.de/tests/15281992)
> > > > [ipxe x86_64 host installation link1](http://10.202.64.26/tests/3121) [link2](http://10.202.96.18/tests/3111)
> > > > [ipxe aarch64 host installation link1](http://10.202.64.26/tests/3119) [link2](http://10.202.96.18/tests/3112)
> > > 
> > > 
> > > I can't open above links... Any hints?
> > 
> > 
> > Finally, I can open them now.
> > Some question about http://10.200.131.12:8666/ipxe/. Hope you can help explain. @waynechen55
> > ```
> > * what is this machine, 10.200.131.12? Do we use it in official test, or switch to official pxe server of OSD?
> > 
> > * I notice that there is no build info for the files under the directory. Where are they from? In official test, will they be updated per build? Which team shall do that?
> > ```
> 
> * The machine does not matter too much. Of course, it should be able to communicate with SUT. It just holds published pxe installation files from [here](https://download.suse.de/ibs/SUSE:/SLFO:/Products:/SL-Micro:/6.1:/ToTest/images/).
> * Because at the moment, these files are only available [here](https://download.suse.de/ibs/SUSE:/SLFO:/Products:/SL-Micro:/6.1:/ToTest/images/), so I need to find a host to host them. When it is available on OSD, test will switch to use resource on OSD. For example, publishing uncompressed [file](https://download.suse.de/ibs/SUSE:/SLFO:/Products:/SL-Micro:/6.1:/ToTest/images/SL-Micro.x86_64-6.1-Default-SelfInstall-Build16.2.install.tar) into OSD repo is the best.
> * Yes, at the moment, these files do not have BUILD info, but the tar file in which they are compressed HAS BUILD info. I am using the latest Build16.2 RC1. Please refer to [this](https://download.suse.de/ibs/SUSE:/SLFO:/Products:/SL-Micro:/6.1:/ToTest/images/SL-Micro.x86_64-6.1-Default-SelfInstall-Build16.2.install.tar).
> 

Sounds good. Yes, when SLE Micro starts to officially support pxe, let's raise the request to infra/tools team for setting up above pxe files.

> I think it is ready to be merged.

Cool! If no more comments raised today, will merge it.

**Review by frankenmichl — APPROVED** [review]

**Comment by alice-suse** [comment]
Thank you all for the review. Merge it now.

**Comment by alice-suse** [comment]
> [successful SL-Micro 6.1 ipxe host installation with kermit (Test run failed partway due to missing needle)](https://openqa.suse.de/tests/15414184)
> [successful SL-Micro 6.1 ipxe host installation with scooter (Test run failed partway due to missing needle)](https://openqa.suse.de/tests/15414186)

@waynechen55  Has the needle MR been opened?

**Comment by Julie-CAO** [comment]
The job group yaml need to be updated accordingly. The test failed in the new build with:

```
YAML_SCHEDULE file not found: 'sle-micro/lib/../schedule/virt_autotest/install_guest_on_slem_kvm_host.yaml'
```

**Comment by czerw** [comment]
To be honest, after I spend more time working with sl micro raw image and baremetal, i would like to ask you revert this PR or remove any virtauto tests specific configuration from ipxe_install. Bare metal installation in this way is not reusable by other teams at all, it is only for virtualization purpose using some undocumented server to get pxe kernel image, which is not versioned at all. Moreover whole things is overcomplicated with many new functions and variables which doesn't make sense. It's my bad that i didn't give feedback, my feedback was mainly about not breaking things, which was ok, but from sl micro and baremetal installation it is just wrong.

**Comment by alice-suse** [comment]
> To be honest, after I spend more time working with sl micro raw image and baremetal, i would like to ask you revert this PR or remove any virtauto tests specific configuration from ipxe_install. 

@czerw  Thanks for your further review and the honest feedback. The author, @waynechen55 , is on vacation until Oct 8. So I'd suggest to hold on reverting until some necessary discussion is made/explained. Maybe some better solution can be found.

Although I may not understand the design as well as Wayne, but let me firstly try to explain something which I saw that maybe not correctly interpreted.


> Bare metal installation in this way is not reusable by other teams at all, it is only for virtualization purpose using some undocumented server to get pxe kernel image, which is not versioned at all. 

SLE Micro pxe boot is quite different from SLES pxe boot, at least from below aspects
- sles get kernel/initrd from pxe(tftp) server, while SLE Micro supports that from http links
- sles supports autoyast installation, while sle micro does not, and instead uses ignition+combustion, which are also supported to be given in pxe boot installation cmdline via giving the http links

So, to extend these within the ipxe_install.pm which in the past supported only "sles", the PR introduces the changes to make it support SLE Micro special pxe installation,  which may make you feel "improper" for "sles". But it is the sacrifice to take if this ipxe_install.pm file would like to have the capability to support both sles and sle micro pxe. And this is reusable for other squads, if they would need sle micro baremetal test. 

Regarding "undocumented server to get pxe kernel image", I guess you mean "MIRROR_HTTP" to get sle micro pxe boot files?  That' s true for the devel verification runs, which just uses a personally setup http server to hold the images and not versioned. The reason is that there is no pxe sle micro support before, and 6.1 has just newly introduced it, so no such versioned images in official image servers.  Meanwhile, the server is configurable, and before official test  deployment on OSD, we are now actually requesting official image server to hold those sle micro pxe files, which will be versioned, see https://progress.opensuse.org/issues/167015. And we will switch to use the versioned official servers once it is ready.

Of course, I agree that this PR's solution is NOT the only way to add capability for sle micro ipxe support, eg introducing another "ipxe_install_xx" for sle micro, can also be a way. But this was not considered at the beginning, and it has been already done in the same file. Let alone that, there will be different pros and cons, for both solutions... 


>Moreover whole things is overcomplicated with many new functions and variables which doesn't make sense. 

I agree, this is another disadvantage of current solution, which may give teams like kernel which do not need sle micro pxe installation pains to maintain the same file... I actually had similar feeling when I firstly saw the changes in the ipxe_install.pm, but after reading more and getting more familiar, I am kind of used to it ;-)

>It's my bad that i didn't give feedback, my feedback was mainly about not breaking things, which was ok, but from sl micro and baremetal installation it is just wrong.

Understand, no worries. You can think about my above reply, and see if some points can be accepted and maybe kind of change your mind. If not, we can open discussions when Wayne is back in Oct.

File: lib/utils.pm [hunk 9]
```diff
@@ -3154,4 +3156,32 @@ sub enable_console_kernel_log {
...
+=head2 is_ipxe_with_disk_image
+
+ is_ipxe_with_disk_image;
+
+Identify whether test runs boots from ipxe and deploy linux disk image built by kiwi or similar programs
+=cut
+
+sub is_ipxe_with_disk_image {
+    return 1 if (is_ipxe_boot and is_disk_image);
+    return 0;
```
Comment (czerw) [inline lib/utils.pm:3184]: I believe this could be  `return (is_ipxe_boot and is_disk_image)` without `if` and double `return`

File: lib/utils.pm [hunk 10]
```diff
@@ -3154,4 +3156,32 @@ sub enable_console_kernel_log {
...
+
+Identify whether test runs with linux disk image built by kiwi or similar programs.
+HDD_1 is usually used if disk image is available on openQA server. Test run attempts
+downloading HDD_1, failure of which leads to failed test run. INSTALL_HDD_IMAGE is
+introduced for installation with disk image which might be located somewhere else
+and also more flexibility.
+=cut
+
+sub is_disk_image {
+    return 1 if ((get_var('HDD_1') or get_var('INSTALL_HDD_IMAGE')) and get_var('BOOT_HDD_IMAGE'));
```
Comment (czerw) [inline lib/utils.pm:3171]: If you have `HDD_1` and `BOOT_HDD_IMAGE` you will always end in installation. Logic should be `(get_var('HDD_1')  && get_var('INSTALL_HDD_IMAGE')` to detect installation scenario.

**Comment by czerw** [comment]
I have created more generic solution for image installation, which needed only minimal change in ipxe_install and works also on other backends  https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20253. It is wtill WIP and works on arm, intel and power is currently in development.

**Comment by czerw** [comment]
Follow for this PR is https://progress.opensuse.org/issues/168850

File: data/virt_autotest/host_unattended_installation_files/combustion/script [hunk 11]
```diff
@@ -5,6 +5,8 @@ grub2_config() {
     file=/etc/default/grub
     sed -i 's/^\s*GRUB_TIMEOUT=.*$/GRUB_TIMEOUT=15/g' $file
     sed -i -r 's/^(\s|#)*GRUB_TERMINAL=.*$/GRUB_TERMINAL="gfxterm console"/g' $file
+    sed -i -r 's/^(\s|#)*GRUB_TERMINAL_INPUT=.*$/GRUB_TERMINAL_INPUT="gfxterm console"/g' $file
+    sed -i -r 's/^(\s|#)*GRUB_TERMINAL_OUTPUT=.*$/GRUB_TERMINAL_OUTPUT="gfxterm console"/g' $file
```
Comment (czerw) [inline data/virt_autotest/host_unattended_installation_files/combustion/script:9]: Congratulation for masking a bug https://bugzilla.suse.com/show_bug.cgi?id=1230844
